compiler/rustc_session/src/config.rs

```diff
@@ -825,6 +825,7 @@ pub enum PrintKind {
     RelocationModels,
     CodeModels,
     TlsModels,
+    TlsDialect,
```

Comment:
Why is this added? Do you have any actual need to know which tls dialect will ne used by rustc on a given target?

Reply:
Also it should probably be made available on nightly only so we can remove it again once the traditional tls dialect is no longer used anywhere.

Reply:
> Why is this added? Do you have any actual need to know which tls dialect will ne used by rustc on a given target?

I followed the example for `TlsModels`, since they seem of equivalent diagnostic value to me. For Android RISC-V targets TLSDESC is the default, and I *think* it may be a requirement.

> Also it should probably be made available on nightly only so we can remove it again once the traditional tls dialect is no longer used anywhere.

I'm not sure that will ever happen. TLSDESC requires support in the dynamic linker, so its imaginable that not all libcs will support that. For instance I've seen embedded targets that implement something like a loader service that support `__tls_get_addr` with overlays, but do not support TLSDESC.  

At any rate, I'm not sure we will ever drop support for traditional TLS in LLVM. It's been discussed, but I think its unlikely until GNU toolchains drop support.

Reply:
> I'm not sure that will ever happen. TLSDESC requires support in the dynamic linker, so its imaginable that not all libcs will support that. 

It can be the default on targets that do support it and not having an option to enable it on targets that don't support it.

Reply:
> > I'm not sure that will ever happen. TLSDESC requires support in the dynamic linker, so its imaginable that not all libcs will support that.
> 
> It can be the default on targets that do support it and not having an option to enable it on targets that don't support it.
> 

I'm not sure how you can make that determination, since you can't know which libc someone is using ahead of time. Being able to configure and opt in/out is necessary, IMO.

> > At any rate, I'm not sure we will ever drop support for traditional TLS in LLVM. It's been discussed, but I think its unlikely until GNU toolchains drop support.
> 
> I mean rustc dropping support for configuring it to anything other than the default for the target.

Failing to expose a code generation option doesn't sound right to me. There are valid reasons someone may want to select a particular variant of an option differently than the default. Until LLVM (and I guess all backends?) drops support altogether, I'm not sure its a good idea for `rustc` to stop exposing that option altogether.

Reply:
> Failing to expose a code generation option doesn't sound right to me. There are valid reasons someone may want to select a particular variant of an option differently than the default. Until LLVM (and I guess all backends?) drops support altogether, I'm not sure its a good idea for rustc to stop exposing that option altogether.

If there is a chance that LLVM ever drops the option, then no, we should not expose the option. We should not intentionally add dead flags that get deprecated to do nothing and lie to our users about their ability to control things. We are not going to become another C compiler that adds 10,000 codegen flags and only tests 1% of them.

Reply:
There will be options to pass arguments directly to codegen backends, and if those suffice, then great, but otherwise we should resist getting into the habit of gaslighting our users by tricking them into thinking they have power and then not being able to fulfill their requests later because of the whims of those codegen backends.

Reply:
I don't think that position is actually sustainable when it comes to codegen options. If we want to expose backend features like code coverage, sanitizers, and profiling, we have to accept that we don't control whether they continue to exist in all backends or continue to work in the same way with all the same knobs as before.

Of course there is always a balance to strike in terms of how much we expose and where the knobs should go. But we should prioritize giving users with minority needs the agency to solve their problems, ideally with tools that either work predictably or give you an error. I'm not sure backend options satisfy those criteria.

It might help to frame this as something more like a [target modifier](https://github.com/rust-lang/rfcs/pull/3716). Support for targets – and by extension, target modifiers – is going to change over time and for different backends. It already does, as we have (correctly) promoted and demoted targets to different tiers based on concerns more myopic to the Rust project, such changing availability of hardware on which to run CI.

Target configurations and codegen options are all about integration across systems: The application, the compiler, the backend, the runtime, the operating system and the hardware it runs on. I do think we should document very clearly that support for this option, or other options like it, depend on the other layers of the stack. Ultimately, users will have to weigh the risk of one of those being pulled out from under them.

In any case, this is for now just an unstable flag, and I think we have precedent for others like it, so I don't think the higher level discussion should block this PR from merging.

Reply:
> I'm not sure how you can make that determination, since you can't know which libc someone is using ahead of time. Being able to configure and opt in/out is necessary, IMO.

Yes, you do. If you use x86_64-unknown-linux-gnu you know for sure that you are using glibc and in particular either the version we document as minimum supported version or newer. Once we document a version that supports TLSDESC as minimum supported version we can make TLSDESC enabled by defaulr for this target. The same applies for all other targets. Every target has exactly one libc implementation that is allowed to be used with it.

Reply:
> Failing to expose a code generation option doesn't sound right to me. There are valid reasons someone may want to select a particular variant of an option differently than the default. 

It would be good to spell those out.

Reply:
> For instance, Android after some particular SDK version may want to drop TLSDESC support, but you may need to retain the ability to target earlier SDKs that predate TLSDESC support in bionic.

Is Android planning to drop support for not using TLSDESC before the last LTS NDK that doesn't support TLSDESC is no longer supported? We only support the most recent LTS NDK.

> Similarly, we know that Fuchsia's driver ABI will only support TLSDESC, but the rest of the system will likely be more flexible about supporting both TLSDESC and traditional TLS.

If Fuchsia has a different ABI between regular userspace and drivers, it would make sense to have separate targets, right? Does anyone care about supporting Fuchsia versions from before TLSDESC got introduced in the first place?

Reply:
> > For instance, Android after some particular SDK version may want to drop TLSDESC support, but you may need to retain the ability to target earlier SDKs that predate TLSDESC support in bionic.
> 
> Is Android planning to drop support for not using TLSDESC before the last LTS NDK that doesn't support TLSDESC is no longer supported? We only support the most recent LTS NDK.
> 

I doubt it, but I'm not an Android maintainer and the precise details are outside of my area of expertise. I mentioned this, since similar concerns were brought up on the LLVM side w.r.t. emulated TLS and the NDK's need to maintain support for it until the NDK versions that used it were no longer supported. Those time horizons were typically long, IIRC. If Rust has a support story here tied to the NDK then maybe the concern I raised is moot.

> > Similarly, we know that Fuchsia's driver ABI will only support TLSDESC, but the rest of the system will likely be more flexible about supporting both TLSDESC and traditional TLS.
> 
> If Fuchsia has a different ABI between regular userspace and drivers, it would make sense to have separate targets, right? 

Drivers in Fuchsia are still user-level programs, and I'm fairly sure we don't want to introduce another triple for them.

>Does anyone care about supporting Fuchsia versions from before TLSDESC got introduced in the first place?

Fuchsia was brought up as an example of a scenario where its useful to have such knobs. I don't think the focus should be on Fuchsia specifically. FWIW, though, we do certainly support older versions of our SDK, though those are usually tied to a specific toolchain and set of libraries.

Reply:
> > It's quite hard to guess what requirements people may have, and failing to expose options that allows them to deviate from the default feels like the wrong approach for a compiler.
> 
> This argument generalizes too much. It could mean that every single decision the compiler could make should have a user-facing option.
> 

There's an argument that it should, even if its not easy to access. Not all software is created equal, and not all compiler decisions are appropriate everywhere. I'm not suggesting that be made the case, just that in many cases not leaving a escape hatch for an optimization leads to worse performance, code size, or a security problem that has to be worked around.

> Compiler flags are part of our user interface. Which means they need testing, we need to consider compatibility, ABI unsoundness, terrible linker errors and so on. So adding one should come with a stronger motivation than "it'd be nice to have a knob, even though there's no immediate need".
> 

I don't view this as "nice to have," I view it as a requirement. Given the history of these options, it seems irrational to fail to give developers the choice of what to do if their needs differ from the default. I'm fine to leave the disagreement there, but I'm not doing this work because projects I support don't need it. Clearly the fact that I'm implementing it suggests some level of need.

> As bjorn says it's possible that this decision is better left to a combination of target specifications and the compiler checking if it's supported (e.g. based on LLVM version) and determining the correct setting based on that.

I disagree. IMO, setting defaults based on the triple is sensible, but locking users in is not, especially when the backend supports both.